File: js/src/gdb/tests/test-jsid.py
# Tests for jsid pretty-printing
# flake8: noqa: F821

assert_subprinter_registered("SpiderMonkey", "JS::PropertyKey")

run_fragment("jsid.simple")

assert_pretty("string_id", '$jsid("moon")')
assert_pretty("int_id", "$jsid(1729)")
unique_symbol_pretty = str(gdb.parse_and_eval("unique_symbol_id")).split("@")[0]
assert_eq(unique_symbol_pretty, '$jsid(Symbol("moon"))')
assert_pretty("registry_symbol_id", '$jsid(Symbol.for("moon"))')
assert_pretty("well_known_symbol_id", "$jsid(Symbol.iterator)")
assert_pretty("void_id", "JS::VoidPropertyKey")

run_fragment("jsid.handles")

assert_pretty("jsid_handle", '$jsid("shovel")')
assert_pretty("mutable_jsid_handle", '$jsid("shovel")')

File: js/src/gdb/tests/test-jsid.cpp
#include "gdb-tests.h"

#include "js/Id.h"  // jsid, JS::PropertyKey
#include "js/String.h"
#include "js/Symbol.h"

FRAGMENT(jsid, simple) {
  const char* chars = "moon";
  JS::Rooted<JSString*> string(cx, JS_NewStringCopyZ(cx, chars));
  JS::Rooted<JSString*> interned(cx, JS_AtomizeAndPinString(cx, chars));
  JS::Rooted<jsid> string_id(cx, JS::PropertyKey::fromPinnedString(interned));
  JS::Rooted<jsid> int_id(cx, JS::PropertyKey::Int(1729));
  JS::Rooted<jsid> unique_symbol_id(
      cx, JS::PropertyKey::Symbol(JS::NewSymbol(cx, interned)));
  JS::Rooted<jsid> registry_symbol_id(
      cx, JS::PropertyKey::Symbol(JS::GetSymbolFor(cx, interned)));
  JS::Rooted<jsid> well_known_symbol_id(
      cx, JS::GetWellKnownSymbolKey(cx, JS::SymbolCode::iterator));
  jsid void_id = JS::PropertyKey::Void();

  breakpoint();

  use(string_id);
  use(int_id);
  use(unique_symbol_id);
  use(registry_symbol_id);
  use(well_known_symbol_id);
  use(void_id);
}

void jsid_handles(JS::Handle<jsid> jsid_handle,
                  JS::MutableHandle<jsid> mutable_jsid_handle) {
  // Prevent the linker from unifying this function with others that are
  // equivalent in machine code but not type.
  fprintf(stderr, "Called " __FILE__ ":jsid_handles\n");
  breakpoint();
}

FRAGMENT(jsid, handles) {
  const char* chars = "shovel";
  JS::Rooted<JSString*> string(cx, JS_NewStringCopyZ(cx, chars));
  JS::Rooted<JSString*> interned(cx, JS_AtomizeAndPinString(cx, chars));
  JS::Rooted<jsid> string_id(cx, JS::PropertyKey::fromPinnedString(interned));
  jsid_handles(string_id, &string_id);
}
